src/javaharness/java/arcs/api/Schema.java

```diff
@@ -0,0 +1,79 @@
+package arcs.api;
+
+import java.util.ArrayList;
+
+public class Schema {
+    enum FieldKind {
+        PRIMITIVE, UNION, TUPLE
+    }
+    public static abstract class Field {
+        public final FieldKind kind;
+        Field(FieldKind kind) {
+            this.kind = kind;
+        }
+        public boolean isPrimitive() { return this.kind == FieldKind.PRIMITIVE; }
+        public boolean isUnion() { return this.kind == FieldKind.UNION; }
+        public boolean isTuple() { return this.kind == FieldKind.TUPLE; }
+    }
+    public static class PrimitiveField extends Field {
+        public final String type;
+        PrimitiveField(String type) {
+            super(FieldKind.PRIMITIVE);
+            this.type = type;
+        }
+    }
+    public static class UnionField extends Field {
+        public final Field types[];
+        public UnionField(Field types[]) {
+            super(FieldKind.UNION);
+            this.types = types;
+        }
+    }
+    public static class TupleField extends Field {
+        public final Field types[];
+        public TupleField(Field types[]) {
+            super(FieldKind.TUPLE);
+            this.types = types;
+        }
+    }
+
+    static Field fieldFromJson(PortableJson json) {
+        String kind = json.getString("kind");
+        if (kind.equals("schema-primitive")) {
+            return new PrimitiveField(json.getString("type"));
+        } else {
+            PortableJson types = json.getObject("types");
+            Field fields[] = new Field[types.getLength()];
```

Comment:
You could do List<Field> fields = new ArrayList<>();, and then fields.add() to push new entries into it.

Reply:
done, thanks!